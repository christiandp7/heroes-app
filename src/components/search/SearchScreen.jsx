import React, { useMemo } from 'react'
import queryString from 'query-string'
import { useLocation } from 'react-router-dom'
// import { heroes } from '../../data/heroes'
import { useForm } from '../../hooks/useForm'
import HeroCard from '../heroes/HeroCard'
import { getHeroesByName } from '../../selectors/getHeroesByName'

const SearchScreen = ({ history }) => {
	const location = useLocation()
	const { q = '' } = queryString.parse(location.search)
	const [formValues, handleInputChange] = useForm({
		searchText: q,
	})

	const { searchText } = formValues

	// const heroesFiltered = getHeroesByName(searchText)
	const heroesFiltered = useMemo(() => getHeroesByName(q), [q])

	const hadleSearch = e => {
		e.preventDefault()
		history.push(`?q=${searchText}`)
	}

	return (
		<>
			<h1>Seach Screen</h1>
			<hr />
			<div className="row">
				<div className="col-5">
					<h4>Seach Form</h4>
					<hr />
					<form onSubmit={hadleSearch}>
						<input
							type="text"
							placeholder="Find your Hero"
							className="form-control"
							name="searchText"
							autoComplete="off"
							value={searchText}
							onChange={handleInputChange}
						/>
						<button
							type="submit"
							className="btn btn-block btn-outline-primary mt-2">
							Search...
						</button>
					</form>
				</div>
				<div className="col-7">
					<h4>Results</h4>
					<hr />
					{q === '' && <div className="alert alert-info">Search a Hero</div>}
					{q !== '' && heroesFiltered.length === 0 && (
						<div className="alert alert-danger">
							There is not a Hero with "{q}"
						</div>
					)}
					<div className="row no-gutters">
						{heroesFiltered.map(hero => (
							<HeroCard key={hero.id} {...hero} />
						))}
					</div>
				</div>
			</div>
		</>
	)
}

export default SearchScreen
