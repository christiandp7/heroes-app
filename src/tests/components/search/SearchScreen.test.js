import { mount } from 'enzyme'
import { MemoryRouter, Route } from 'react-router-dom'
import SearchScreen from '../../../components/search/SearchScreen'

describe('<SearchScreen /> test', () => {
	test('should be shown with right values', () => {
		const wrapper = mount(
			<MemoryRouter initialEntries={['/search']}>
				<Route path="/search" component={SearchScreen} />
			</MemoryRouter>,
		)
		expect(wrapper).toMatchSnapshot()
		expect(wrapper.find('.alert-info').text().trim()).toBe('Search a Hero')
	})

	test('should show Batman and the input with the same queryString value', () => {
		const wrapper = mount(
			<MemoryRouter initialEntries={['/search?q=batman']}>
				<Route path="/search" component={SearchScreen} />
			</MemoryRouter>,
		)

		expect(wrapper.find('input').prop('value')).toBe('batman')
		expect(wrapper).toMatchSnapshot()
	})

	test('should show a error message if Hero does not exists', () => {
		// dsd
		const wrapper = mount(
			<MemoryRouter initialEntries={['/search?q=nohero']}>
				<Route path="/search" component={SearchScreen} />
			</MemoryRouter>,
		)

		expect(wrapper.find('.alert-danger').exists()).toBe(true)
		expect(wrapper).toMatchSnapshot()
	})

	test('should call PUSH from history', () => {
		const history = {
			push: jest.fn(),
		}
		const wrapper = mount(
			<MemoryRouter initialEntries={['/search?q=nohero']}>
				<Route
					path="/search"
					component={() => <SearchScreen history={history} />}
				/>
			</MemoryRouter>,
		)

		wrapper.find('input').simulate('change', {
			target: {
				name: 'searchText',
				value: 'batman',
			},
		})

		wrapper.find('form').prop('onSubmit')({ preventDefault() {} })

		expect(history.push).toHaveBeenCalledWith(`?q=batman`)
	})
})
