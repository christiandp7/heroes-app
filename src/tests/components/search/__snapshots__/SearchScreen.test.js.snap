// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<SearchScreen /> test should be shown with right values 1`] = `
Array [
  <h1>
    Seach Screen
  </h1>,
  <hr />,
  <div
    className="row"
  >
    <div
      className="col-5"
    >
      <h4>
        Seach Form
      </h4>
      <hr />
      <form
        onSubmit={[Function]}
      >
        <input
          autoComplete="off"
          className="form-control"
          name="searchText"
          onChange={[Function]}
          placeholder="Find your Hero"
          type="text"
          value=""
        />
        <button
          className="btn btn-block btn-outline-primary mt-2"
          type="submit"
        >
          Search...
        </button>
      </form>
    </div>
    <div
      className="col-7"
    >
      <h4>
        Results
      </h4>
      <hr />
      <div
        className="alert alert-info"
      >
        Search a Hero
      </div>
      <div
        className="row no-gutters"
      />
    </div>
  </div>,
]
`;

exports[`<SearchScreen /> test should show Batman and the input with the same queryString value 1`] = `
Array [
  <h1>
    Seach Screen
  </h1>,
  <hr />,
  <div
    className="row"
  >
    <div
      className="col-5"
    >
      <h4>
        Seach Form
      </h4>
      <hr />
      <form
        onSubmit={[Function]}
      >
        <input
          autoComplete="off"
          className="form-control"
          name="searchText"
          onChange={[Function]}
          placeholder="Find your Hero"
          type="text"
          value="batman"
        />
        <button
          className="btn btn-block btn-outline-primary mt-2"
          type="submit"
        >
          Search...
        </button>
      </form>
    </div>
    <div
      className="col-7"
    >
      <h4>
        Results
      </h4>
      <hr />
      <div
        className="row no-gutters"
      >
        <div
          className="col-sm-6 col-md-6"
        >
          <div
            className="card"
          >
            <div
              className="row no-gutters"
            >
              <div
                className="col-sm-12 col-md-4"
              >
                <img
                  alt="Batman"
                  className="img-fluid rounded-start"
                  src="assets/heroes/dc-batman.jpg"
                />
              </div>
              <div
                className="col-sm-12 col-md-8"
              >
                <div
                  className="card-body"
                >
                  <h5
                    className="card-title"
                  >
                    Batman
                  </h5>
                  <p
                    className="card-text"
                  >
                    Bruce Wayne
                  </p>
                  <p
                    className="card-text"
                  >
                    <small
                      className="text-muted"
                    >
                      Detective Comics #27
                    </small>
                  </p>
                  <a
                    className="btn btn-primary"
                    href="/hero/dc-batman"
                    onClick={[Function]}
                  >
                    Mas...
                  </a>
                </div>
              </div>
            </div>
          </div>
        </div>
      </div>
    </div>
  </div>,
]
`;

exports[`<SearchScreen /> test should show a error message if Hero does not exists 1`] = `
Array [
  <h1>
    Seach Screen
  </h1>,
  <hr />,
  <div
    className="row"
  >
    <div
      className="col-5"
    >
      <h4>
        Seach Form
      </h4>
      <hr />
      <form
        onSubmit={[Function]}
      >
        <input
          autoComplete="off"
          className="form-control"
          name="searchText"
          onChange={[Function]}
          placeholder="Find your Hero"
          type="text"
          value="nohero"
        />
        <button
          className="btn btn-block btn-outline-primary mt-2"
          type="submit"
        >
          Search...
        </button>
      </form>
    </div>
    <div
      className="col-7"
    >
      <h4>
        Results
      </h4>
      <hr />
      <div
        className="alert alert-danger"
      >
        There is not a Hero with "
        nohero
        "
      </div>
      <div
        className="row no-gutters"
      />
    </div>
  </div>,
]
`;
